### DISCUSSION TITLE
Why `validateOnly` need to get all data?

### DISCUSSION THREAD

**Discussion by chuoke** [livewire/livewire#4823] [1. Help] [answered]
There are some validation rules  are shared out of livewire component. In some case, in this livewire component logic no need some fields. But the following code prevents the shared rules from being used

https://github.com/livewire/livewire/blob/a7cccf9ae4f8fe46886efe10ca491a6a6f4e7e2f/src/ComponentConcerns/ValidatesInput.php#L232
It will throw a exception as : `No property found for validation: [name]`.

Can change to like this?
```php
$data = $this->getDataForValidation($rulesForField);
```

**Comment by joshhanley** [comment]
This issue has been converted to a discussion as we are now using discussions to track bugs and ideas.
See this announcement for more details https://github.com/livewire/livewire/discussions/3147

**Reply by chuoke** [reply]
Sorry, not know that.First in.

**Reply by joshhanley** [reply]
@chuoke no worries 🙂 how did you manage to create an issue though? As our "New Issue" button on the issues page should take you to a page to create a new discussions instead of an issue.

**Reply by chuoke** [reply]
First I'm not log in. When I click the Issue button be asked to log in. After logging in redirected to the create issue page.

**Comment by joshhanley** [comment]
@chuoke in regards to your issue though, I'm not sure I follow. Can you share your component code (properties and rules) and explain what you mean?

Thanks!

**Comment by chuoke** [comment]
Add my code

The shared rules:

```php
trait UserCreateRules
{
    use PasswordRules;

    /**
     * Validate data of register user.
     *
     * @param  array  $input
     * @return array
     */
    public function userCreateRules()
    {
        return [
            'name' => ['string', 'max:20'],
            'username' => ['required', 'string', 'max:25', 'alpha_dash', Rule::unique(User::class),],
            'email' => [ 'required', 'string',  'email',  'max:40', Rule::unique(User::class),  ],
            'password' => $this->passwordRules(),
        ];
    }
}
```

Livewire component:

```php
class Register extends Component
{
    use UserCreateRules;

    public $username;
    public $email;
    public $password;

    public function mount()
    {
        $this->previous = URL::previous();
    }

    public function render()
    {
        return view('livewire.auth.register');
    }

    public function updated($propertyName)
    {
        $this->validateOnly($propertyName, $this->userCreateRules());
    }

    public function submit()
    {
        event(new Registered($user = (app(CreatesNewUsers::class))->create([
            'username' => $this->username,
            'email' => $this->email,
            'password' => $this->password,
        ])));

        app(StatefulGuard::class)->login($user, true);

        return \redirect($this->previous);
    }
}
```

As you see the component has no `name` property. When some change, It will throw an exception as: `No property found for validation: [name]` from `$this->validateOnly($propertyName, $this->userCreateRules());`.

**Comment by joshhanley** [comment, marked as answer]
@chuoke ah ok, I understand.

The reason `validateOnly` needs all data is because there are validation rules that require other data to work like `if` or `exclude_if`, etc.

For your example, I would be taking the rules and filtering them down, so there is no name rule if there is no name property on the component. Personally I don't like sharing rules like that, I think rules should be setup on the component to match the properties in the component.

Hope this helps!

**Reply by chuoke** [reply]
> The reason `validateOnly` needs all data is because there are validation rules that require other data to work like `if` or `exclude_if`, etc.

oh, 😂. Be enlightened. Thanks!
